Normalize specified types before intersection

herndlm opened the pull request:
Closes https://github.com/phpstan/phpstan/issues/6329

This PR is improving type specification with combined BooleanAnd and BooleanIf expressions in 2 ways. Therefore the title does not match anymore :) I think something like "Improve type specification for combined && and || expressions" would fit better.

1. It normalizes the SpecifiedTypes by removing the sureNotTypes from the relevant sureTypes which results in e.g. `AccessoryNonEmptyStringType` or `NonEmptyArrayType`. Without this, sureNotTypes might not be considered when intersecting instances of `SpecifiedType`, which leads to type information loss.

<details>
<summary>Details about the normalization</summary>

In order to narrow down the variable `$a` from `mixed ` to `non-empty-string|null` the following code could (it's maybe a bit odd but this is not relevant) be used in an if (truthy context):
```php
is_string($a) && '' !== $a || null === $a
```
This can be described with the expression
```php
new Expr\BinaryOp\BooleanOr(
	new Expr\BinaryOp\BooleanAnd(
		new FuncCall(new Name('is_string'), [new Arg(new Variable('a'))])
		new NotIdentical(new String_(''), new Variable('a')),
	),
	new Identical(new Expr\ConstFetch(new Name('null')), new Variable('a')),
)
```

The interesting parts are the inner BooleanAnd and the outer BooleanOr. If the TypeSpecifier looks at this it will basically recursively determine the `SpecifiedTypes`. Without going into too much details, that the relevant people know better than me anyway, the following will be the simplified relevant sureTypes and sureNotTypes for `$a` of the relevant SpecifiedTypes instances:

### BooleanAnd
left expression: `StringType` sureType
right expression:  `ConstantStringType` with value `''` sureNotType

#### Result
BooleanAnd uses `SpecifiedTypes->unionWith` for a truthy context like here which results in a new SpecifiedTypes having both of those

### BooleanOr
left expression: the above union => `StringType` sureType,  `ConstantStringType` with value `''` sureNotType
right expression: `NullType` sureType

#### Result
=> BooleanOr uses `SpecifiedTypes->intersectWith`

##### Before
intersectWith only looks at the sureTypes and sureNotTypes of both SpecifiedTypes instances if their expression string occurs also in both. Meaning, it will union (for a truthy context like here) the `StringType` and `NullType` because both SpecifiedTypes instances have sureTypes for `$a`. But, it will not use the sureNotType if there is not a sureNotType in the other SpecifiedTypes for `$a`. Therefore it results in a SpecifiedTypes instance with a `StringType|NullType` sureTypes union only and basically looses the sureNotType information that is relevant for the left expression.

##### After
both SpecifiedTypes instances are normalized which "moves" the sureNotType into the sureType resulting in a `StringType&AccessoryNonEmptyStringType` intersection for the left expression basically. The other SpecifiedTypes instance is not affected here by normalization. The intersectWith logic then unionizes (for a truthy context like here) both instances resulting in a `(StringType&AccessoryNonEmptyStringType)|NullType` union sureType and therfore does not loose the sureNotType information which is coupled to the left expression but does not make any sense for the right one.

Without the BooleanOr the sureNotType is not lost and the MutatingScope can use it to determine that the outcome is `non-empty-string`. With BooleanOr we have to normalize this already here IMO before intersecting in order to not loose the information.

</details>

2. It makes use of the `MutatingScope` in `BooleanAnd` and `BooleanOr` expressions to filter by the specified types of the left espression and uses the resulting mutated scope in the right expression. This ensures that information gathered on the left side can be used on the right side. See https://github.com/phpstan/phpstan-src/pull/1016#issuecomment-1040372045

herndlm commented on src/Analyser/TypeSpecifier.php:
```diff
@@ -482,6 +482,8 @@ public function specifyTypesInCondition(
 					if ($argType->isArray()->yes()) {
 						$result = $result->unionWith($this->create($expr->right->getArgs()[0]->value, new NonEmptyArrayType(), $context, false, $scope));
+					} else {
+						$result = $result->unionWith($this->create($expr->right->getArgs()[0]->value, new ConstantArrayType([], []), $context->negate(), false, $scope));
```
This fixes `is_string($a) && strlen($a) > 0 || null === $a`. The problem seems to be that in that case `$argType` is still `MixedType` here. I'm not sure why, but this smells a bit like a hacky workaround? If so, I'm fine with removing it and the relevant assertions in order to fix the rest for now. I was not successful debugging everything that is going on leading to `MutatingScope::$resolvedTypes` being set incorrectly..

herndlm replied:
same thing as a couple lines further down but for `is_array($a) && count($a) > 0 || null === $a`

herndlm commented on src/Analyser/TypeSpecifier.php:
```diff
@@ -501,6 +503,8 @@ public function specifyTypesInCondition(
 					if ($argType instanceof StringType) {
 						$result = $result->unionWith($this->create($expr->right->getArgs()[0]->value, new AccessoryNonEmptyStringType(), $context, false, $scope));
+					} else {
+						$result = $result->unionWith($this->create($expr->right->getArgs()[0]->value, new ConstantStringType(''), $context->negate(), false, $scope));
```
This fixes `is_string($a) && strlen($a) > 0 || null === $a`. The problem seems to be that in that case `$argType` is still `MixedType` here. I'm not sure why, but this smells a bit like a hacky workaround? If so, I'm fine with removing it and the relevant assertions in order to fix the rest for now. I was not successful debugging everything that is going on leading to `MutatingScope::$resolvedTypes` being set incorrectly..

herndlm replied:
I think there are a couple of interesting parts here and in #1012 
- the type specification for the left part for `is_string` is coming from a TypeSpecifyingExtension
- the right part does not know about the type from the left part and if it asks the scope for it, it gets mixed. this is sometimes a problem, sometimes not. it looks though like here it can still find a type for `strlen($a) > 0` in the right part
![image](https://user-images.githubusercontent.com/5738896/153940688-7badb8a6-df1e-4bbb-9d5f-90500b779be2.png)
the scope is using `filterBySpecifiedTypes` afterwards, and then has the right type (non-empty-string) if asked for the expression or the variable type
- but in this case, there is also the BooleanOr and this is where, again, type information (the `strlen($a) > 0`) is lost via intersection of 2 SpecifiedTypes instances

herndlm replied:
OK, I came up with something, resolving this for now and adding a new comment down below

staabm commented on tests/PHPStan/Analyser/data/bug-6329.php:
```diff
@@ -0,0 +1,160 @@
+ */
+function nonEmptyArray1($a): void
+{
+	if (is_array($a) && [] !== $a || null === $a) {
```
This makes me think whether this also works for e.g.
`is_array($a) && [] !== $a || true === $a`
`is_array($a) && [] !== $a || false === $a`
`is_array($a) && [] !== $a || “abc“ === $a`
`is_array($a) && [] !== $a || 123 === $a`

But lets wait for ondrejs feedback first. The PR as is seems to add already a lot of great type improvements

herndlm replied:
I did a quick test, all of those cases are working fine. Which is not surprising I guess :)
Results are
```
non-empty-array|true
non-empty-array|false
'abc'|non-empty-array
123|non-empty-array
```
I did not check though if those were already working before or not :)

staabm commented on tests/PHPStan/Analyser/TypeSpecifierTest.php:
```diff
@@ -965,6 +965,45 @@ public function dataCondition(): array
+					new Expr\BinaryOp\BooleanAnd(
+						$this->createFunctionCall('is_array', 'a'),
+						new Expr\BinaryOp\Greater(
+							$this->createFunctionCall('count', 'a'),
```
where count() is supported, at best also sizeof() would work

herndlm replied:
yeah, definitely. It's handled by the exact same block of code that handles `count` and `strlen` too. I would therefore not add another test if that's fine.

herndlm commented:
The failing tests are either unrelated or, in case of webmozart-assert, actual fixes

Marking this as ready, but I have the feeling the 2 typespecifier workarounds should be reverted for now, just let me know

ondrejmirtes commented:
Hi, thank you for the research. Can you please describe in more detail how does SpecifiedTypes look before and after the normalization and what does it achieve further down the line? Thanks.

herndlm commented:
@ondrejmirtes I wrote a short novel hidden behind the Details section. I hope this explains everything.

ondrejmirtes commented:
Perfect explanation, I now understand the problem :) I'm gonna let it sit for a while to think about it, maybe I'll come up even with a better solution. What I'm worried about a bit is a falsy context where the meaning of sureTypes/sureNotTypes is switched and it should be doing the opposite I think. Anyway, it's now my turn to do something about it, thank you.

herndlm commented:
I added another commit and with that one I managed to
- get rid of my hacky workarounds for `strlen` and `count` that I added in the first commit
- fix what I described in #1012

It makes use of the `MutatingScope` when handling `BooleanAnd` and `BooleanOr` by filtering with the left expression. And that mutated scope is used for the right expression. I think this was needed because the left expression could narrow down a `mixed` variable via e.g. `is_string` to, well, `string`. And that information is currently not used in the right expression which can make problems with some type extensions that ask the scope via `getType` and use that information internally.
I think this change makes PHPStan understand some complex true and false context conditionals that it did not understand before 🎉 At least if I'm not completely mistaken. But I'll think about it a bit more. Maybe it also makes sense to add more complex test cases somewhere

ondrejmirtes commented:
Are any of the new test cases and bugfixes about my concern?

> What I'm worried about a bit is a falsy context where the meaning of sureTypes/sureNotTypes is switched and it should be doing the opposite I think.

herndlm commented:
> Are any of the new test cases and bugfixes about my concern?

There were already test cases that check the type in the else statement, I added some more, so yes :) And AFAIK and could see via debugger, it uses a falsy context then to create the specified types and filter by them in the scope. I think this should be covered.

ondrejmirtes commented:
Alright, I'm gonna merge it :) Please send a fix to webmozart-assert extension afterwards, thanks.

ondrejmirtes commented:
Thank you!

herndlm commented:
Finally I can sleep tonight xD But serious, this kept me thinking, busy and step debugging a couple of hours

> Alright, I'm gonna merge it :) Please send a fix to webmozart-assert extension afterwards, thanks.

In order to do that I need a new phpstan release to specify as minimum supported version, I think. Shall I comment-out the two problematic lines for now and add them back in correctly after a new phpstan version was tagged?

ondrejmirtes commented:
Also, feel free to continue thinking about it and have it in the back of your head, it's often that people come up with improvements afterwards and I love many little PRs :) Generics were implemented over dozens of small PRs :)

staabm commented:
great job @herndlm !
